junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TimeoutDurationParser.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.extension;
+
+import static java.util.concurrent.TimeUnit.DAYS;
+import static java.util.concurrent.TimeUnit.HOURS;
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.regex.Pattern.CASE_INSENSITIVE;
+import static java.util.regex.Pattern.UNICODE_CASE;
+
+import java.time.format.DateTimeParseException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @since 5.5
+ */
+class TimeoutDurationParser {
+
+	private static final Pattern PATTERN = Pattern.compile("([1-9]\\d*) ?((?:[nμm]?s)|m|h|d)?",
+		CASE_INSENSITIVE | UNICODE_CASE);
+	private static final Map<String, TimeUnit> UNITS_BY_ABBREVIATION;
+
+	static {
+		Map<String, TimeUnit> unitsByAbbreviation = new HashMap<>();
+		unitsByAbbreviation.put("ns", NANOSECONDS);
+		unitsByAbbreviation.put("μs", MICROSECONDS);
```

Comment:
Now that I've thought about it, do we really expect a timeout with the granularity of microseconds or nanoseconds to be meaningful in a test suite, especially if tests are executed in parallel?

I can't really imagine anything more fine-grained than milliseconds being robust in a test suite.

Reply:
Probably not, I just included it for consistency with the annotation. Or do you think we should define our own enum?

At least `TimeUnit` does not have a `FOREVER` (unlike `ChronoUnit`)... 😏

Reply:
> Probably not, I just included it for consistency with the annotation. 

Understood.

> Or do you think we should define our own enum?

Good question. I'm not sure that "days" makes sense either. "hours" is already pushing it. 

So maybe it would provide a better user experience if users could only select time units that are reasonable for the use case at hand. For example, we could limit it to the following and internally map to `TimeUnit` enum constants for those.

- milliseconds
- seconds
- minutes
- hours

WDYT?

> At least `TimeUnit` does not have a `FOREVER` (unlike `ChronoUnit`)... 😏

Indeed.

Reply:
Let's stick with `TimeUnit` "as-is". It is used with a `long` through-out the Java SE API as well: https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/util/concurrent/class-use/TimeUnit.html

Reply:
I agree with @sormuras on this one.

@sbrannen Ok to leave "as-is" or would you want to change the parser?

Reply:
I think it's OK to stick with `TimeUnit` as the enum.

As for changing the parser, I have mixed feelings. TBH, I don't think anyone will ever have a test timeout in days, but as for whether or not it will ever make sense to support microseconds or nanoseconds... I cannot predict the future.

So maybe it's OK to just support all `TimeUnit` constants as in the status quo.

One option would be to provide a _tip_ in the Javadoc and User Guide pointing out that one should pick reasonable time units, but then again... maybe that's too self explanatory. 😇 

Long story, short: I'll leave it up to you.